azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,108 @@
+import queue
+import logging
+from threading import Event
+
+logger = logging.getLogger(__name__)
+
+
+class MessageQueue(queue.Queue):
+    def __init__(self, on_enable, on_disable):
+        super(MessageQueue, self).__init__()
+        self._enabled = False
+        self._on_enable = on_enable
+        self._on_disable = on_disable
+
+    def enable(self):
+        if not self._enabled:
+            self._on_enable()  # must do this before changing status!
+            self._enabled = True
+
+    def disable(self):
+        if self._enabled:
+            self._enabled = False
+            self._on_disable()  # must do this after changing status!
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, v):
+        if v:
+            self.enable()
+        else:
+            self.disable()
+
+
+class MessageQueueManager(object):
+    def __init__(self, enable_feature_fn, disable_feature_fn):
+        self.enable_feature = enable_feature_fn
+        self.disable_feature = disable_feature_fn
+        self.c2d_message_queue = MessageQueue(
+            self._c2d_message_queue_on_enable, self._c2d_message_queue_on_disable
+        )  # not yet enabled
+        self.input_message_queues = {}
+
+    def _input_message_queue_on_enable(self):
+        # Enable input messages if no input queues are enabled
+        if not any(queue.enabled for queue in self.input_message_queues.values()):
+            self.enable_feature("input")
+
+    def _c2d_message_queue_on_enable(self):
+        self.enable_feature("c2d")
+
+    def _input_message_queue_on_disable(self):
+        # Disable input messages as a feature if all input queues are disabled
+        if all(not queue.enabled for queue in self.input_message_queues.values()):
+            self.disable_feature("input")
+
+    def _c2d_message_queue_on_disable(self):
+        self.disable_feature("c2d")
+
+    def get_input_message_queue(self, input_name):
+        try:
+            queue = self.input_message_queues[input_name]
+        except KeyError:
+            # Create new MessageQueue for input if it does not yet exist
+            queue = MessageQueue(
+                self._input_message_queue_on_enable, self._input_message_queue_on_disable
+            )
+            self.input_message_queues[input_name] = queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def get_c2d_message_queue(self):
+        queue = self.c2d_message_queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def receive_input_message(self, input_name, message_received):
+        delivered = False
```

Comment:
>delivered = False [](start = 8, length = 17)

Is it a convention to return Boolean variables ? can we not return true from the if and otherwise by default return false ?

Reply:
I'd be okay with that, but generally the IoT team has preferred one exit methods, i.e. only one return statement, so I did it this way.

---
In reply to: [259134718](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r259134718) [](ancestors = 259134718)